api-get-object.go

```diff
@@ -127,9 +127,6 @@ func (c *Client) GetObject(ctx context.Context, bucketName, objectName string, o
 				} else {
 					// First request is a Stat or Seek call.
 					// Only need to run a StatObject until an actual Read or ReadAt request comes through.
-
-					// Remove range header if already set, for stat Operations to get original file size.
-					delete(opts.headers, "Range")
```

Comment:
Does setting range change the stat output ? Let's get @harshavardhana  review here. But I agree, this is not the right place to delete a header here

Reply:
See linked bug. It creates a weird inconsistency where the Stat response changes based on whether you have read from the returned object or not.

Reply:
> See linked bug. It creates a weird inconsistency where the Stat response changes based on whether you have read from the returned object or not.

Yes, I agree, in general it is just wrong to remove a header at this location. But my question is why the code was there in the first place, so just asking for @harshavardhana input

Reply:
This code was there to ensure that we always store the original file size not the size of the range.

We only need to honor range for the body when 'reading' but Stat() output was using the total size to ensure Seek() can return appropriate errors.

If we honor Stat() to honor the range required for the body then the problems might show up with Seek().

Reply:
Remember Stat() is like (*os.File).Stat() which returns the original size of the object not the range you asked. 

The range is basically here given for situations where the caller simply doesn't care about Stat() or Seek()

However users like mc don't really use such low level usage. It relies on Seek() to have correct behavior and a Seek() must not change the Stat() result this is why the range header is removed here. Since we are interested in the original size of the object.

Reply:
@harshavardhana Then calling Stat after reading cannot reuse the the data info from GetObject should be changed to always do HeadObject if there is a range.

Reply:
Correct we should fix that @klauspost it must be "independent" of any existing state.

Reply:
@r-scheele Let's do that then. If a range was set and we fetched some data, discard the information we already got and perform a clean HeadObject.

Reply:
Alright @klauspost, I'll look into it.